### PR TITLE
Fix Regression in Client-side Interceptors

### PR DESCRIPTION
**Pull request by gnossen**
https://github.com/grpc/grpc/pull/20812 caused a regression for unary-unary client interceptors in the case of an exception. This PR fixes that regression and adds tests for error conditions with client-side interceptors.

### CODE REVIEW

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -263,35 +263,45 @@ def _rpc_state_string(class_name, rpc_state):
                 rpc_state.debug_error_string)
 
 
-class _RpcError(grpc.RpcError, grpc.Call):
+class _RpcError(grpc.RpcError, grpc.Call, grpc.Future):
     """An RPC error not tied to the execution of a particular RPC.
 
+    The state passed to _RpcError must be guaranteed not to be accessed by any
+    other threads.
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_channel.py]: Can we use the copy-on-creation mode here? So the `RpcState` object being copied in the initialization function? In my mind, `_RpcError` is a data structure to store error info, it should be fine to keep a snapshot of the `RpcState`.

File: src/python/grpcio/grpc/_channel.py [hunk 2]
```diff
@@ -263,35 +263,45 @@ def _rpc_state_string(class_name, rpc_state):
...
+        if state.cancelled:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for a cancelled RPC.")
+        if state.code is grpc.StatusCode.OK:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for a successfully completed RPC."
+            )
+        if state.code is None:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for an incomplete RPC.")
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_channel.py]: If we are copying state, maybe we can map above exception cases into corresponding `code` and `details`? E.g. if the `code` is `None` during initialization, maybe we can assign `UNKNOWN` to it.

File: src/python/grpcio/grpc/_channel.py [hunk 3]
```diff
@@ -302,6 +312,41 @@ def __repr__(self):
     def __str__(self):
         return self._repr()
 
+    def cancel(self):
+        """See grpc.Future.cancel."""
+        return False
+
+    def cancelled(self):
+        """See grpc.Future.cancelled."""
+        return False
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_channel.py]: Should this function return the cancellation state of the `RpcError` or the RPC call? Apply to other methods.

File: src/python/grpcio/grpc/_channel.py [hunk 4]
```diff
@@ -302,6 +312,41 @@ def __repr__(self):
...
+        return self
+
+    def traceback(self, timeout=None):  # pylint: disable=unused-argument
+        """See grpc.Future.traceback."""
+        try:
+            raise self
+        except grpc.RpcError:
+            return sys.exc_info()[2]
+
+    def add_done_callback(self, fn, timeout=None):  # pylint: disable=unused-argument
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_channel.py]: nit: for `pylint: disable=unused-argument`, can we name it `unused_timeout` to suppress the warning?

File: src/python/grpcio_tests/tests/unit/_interceptor_test.py [hunk 5]
```diff
@@ -232,7 +242,14 @@ def intercept_service(self, continuation, handler_call_details):
...
         self.record.append(self.tag + ':intercept_unary_unary')
-        return continuation(client_call_details, request)
+        result = continuation(client_call_details, request)
+        assert isinstance(
+            result,
+            grpc.Call), '{} is not an instance of grpc.Call'.format(result)
+        assert isinstance(
+            result,
+            grpc.Future), '{} is not an instance of grpc.Future'.format(result)
+        return result
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_interceptor_test.py]: optional: should we log `type(result)` as well, might help ourselves in future.

File: src/python/grpcio_tests/tests/unit/_interceptor_test.py [hunk 6]
```diff
@@ -440,6 +464,24 @@ def testInterceptedUnaryRequestBlockingUnaryResponse(self):
...
+                                             'c1', self._record),
+                                         _LoggingInterceptor(
+                                             'c2', self._record))
+
+        multi_callable = _unary_unary_multi_callable(channel)
+        with self.assertRaises(grpc.RpcError):
+            multi_callable(
+                request,
+                metadata=(('test',
+                           'InterceptedUnaryRequestBlockingUnaryResponse'),))
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_interceptor_test.py]: Should we check the content of the `grpc.RpcError`?

File: src/python/grpcio_tests/tests/unit/_interceptor_test.py [hunk 7]
```diff
@@ -70,9 +72,13 @@ def handle_unary_unary(self, request, servicer_context):
                 'testkey',
                 'testvalue',
             ),))
+        if request == _EXCEPTION_REQUEST:
+            raise RuntimeError()
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_interceptor_test.py]: It might be helpful to raise a special exception private to this test case, so the test case can deterministic catch it. In case, we somehow perform some change that causes `RuntimeError`.

File: src/python/grpcio/grpc/_channel.py (hunk 4 above)
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: Unfortunately, that would mean that we don't actually fulfill the `grpc.Future` interface.

File: src/python/grpcio/grpc/_channel.py (hunk 2 above)
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: I don't see the advantages to this. Can you please elaborate?

Comment (lidizheng) [reply src/python/grpcio/grpc/_channel.py]: Since all tests are passing, I guess I'm talking about some corner cases here.

For the semantic of the `_RpcError` class, IMO it is responsible to construct an informative exception to propagate to application.

In cases that needs to raise an exception, we should prioritize the propagate of any debug details, and make sure the construction of the exception do not fail. If the initialization raises an exception, the original exception will be lost.

So, for cancelled RPC and incomplete RPC, I think it is fine to allow them to be `grpc.RpcError`. For successfully completed RPC, I agree we should raise an `ValueError` since that might indicate logic error in our framework.

Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: If these `ValueError`s were ever actually raised, it would be indicative of a bug in our own code. Now that I've seen the test suite pass with these checks in, I would rather just remove the checks completely.

**Comment by gnossen** [comment]
@lidizheng PTALA

**Comment by gnossen** [comment]
@lidizheng The main difference for `_RpcError` is not whether it works on the single-threaded or multi-threaded path, but rather whether it serves an *in-progress* RPC or not. Note that the docstring already explains this. This is what allows the implementations of all of the methods to be so simple.

File: src/python/grpcio/grpc/_channel.py (hunk 2 above)
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: `ValueError`s removed.

File: src/python/grpcio/grpc/_channel.py (hunk 3 above)
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: Discussed offline. This class is supposed to have a very limited scope. The name has been changed to `_InactiveRpcError` to indicate this.

**Comment by gnossen** [comment]
Unrelated ObjC failure.
